spec.md

```diff
@@ -0,0 +1,276 @@
+
+### DELETE `/ipfs/<CID>`
+
+Unpin the provided CID.
+
+#### Response
+200 OK if successful.
+
+
+## IPNS APIs
+
+
+### GET or HEAD `/ipns/<key/domain>[/<path>]`
+
+#### Options
+Same as GET `/ipfs/<CID>[/<path>]`
+
+#### Response
+Resolves the IPNS key to an IPFS path, and then returns the same response as `/ipfs/`. Also supports resolving custom domains with [DNSLink](https://dnslink.io/).
+
+The header `X-IPFS-Path` is set to the URL path. `IPFS-Hash` is set to CID of the IPFS path provided, this may differ from the CID in the URL path. This allows for getting the CID of an IPNS path/key.
+
+### POST `/ipns/<key>[/<path>]`
+
+Update all or part of the content behind an IPNS name. The new content is pinned, and the previous content is unpinned if it exists.
+
+#### Request Body
+A valid IPFS path in the format `/ipfs/<CID>[/<path>]` A naked CID is also valid.
+
+#### Response
+A 301 Temporary Redirect to `ipns://<key>[/<path>]`. The header `X-IPNS-Path` is set to the `/ipns/<key>[/<path>]`.
+
+### POST `/ipns/localhost[/<path>]`
+
+Behaves the same as `/ipns/<key>[/<path>]`.
+
+#### Query params
+- `key=some_name`: **Required**. This specifies the key you want by a name instead of by public key. If a key with this name doesn't exist, it will be created.
+
+#### Response
+The same as `/ipns/<key>[/<path>]`. Note this allows for extraction of the actual public key from future use and link sharing.
+
+### PUT `/ipns/<key>[/<path>]`
+
+The same as POST `/ipns/<key>[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### PUT `/ipns/localhost[/<path>]`
+
+The same as POST `/ipns/localhost[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### DELETE `/ipns/<key>[/<path>]`
+
+Remove content behind an IPNS name. The new content is pinned, and the previous content is unpinned.
+
+#### Response
+A 301 Temporary Redirect to `ipns://<key>[/<path>]`. The path in the redirect is the directory containing the deletion path. So if the path in the URL was `/ipns/<key>/path/to/file.ext`, the redirect will be to `ipns://<key>/path/to`.
+
+The header `X-IPNS-Path` is set to `/ipns/<key>[/<path>]`, with the same path logic as above.
+
+### DELETE `/ipns/localhost[/<path>]`
+
+Behaves the same as `/ipns/<key>[/<path>]`.
+
+#### Query params
+- `key=some_name`: **Required**. This specifies the key you want by a name instead of by public key. If a key with this name doesn't exist, it will be created.
+
+#### Response
+The same as `/ipns/<key>[/<path>]`. Note this allows for extraction of the actual public key from future use and link sharing.
+
+## Pubsub
+
+An experimental pubsub API is supported to give users direct access to sending and receiving pubsub messages.
```

Comment:
Should we also expect the `Accept` header to be specified to accept event-streams? I've been using that in other protocols so that you can still GET some data as JSON if it's useful.

E.g. `GET /pubsub/<topic> {Accept:application/json}` could get back some info about the subscribe state and your own ID or something.

Reply:
Sure if you think there's some useful info we can get back we could do that. Could be good for debugging also. Do you want to send a schema here or in a new issue or something? Then I'll add it.

Also I'm thinking `Accept: text/event-stream` will stream data, and any other header will return JSON.

Reply:
> Do you want to send a schema here or in a new issue or something?

Like a JSON Schema?

```json
{
  "type": "object",
  "required": ["topic", "id", "subscribed", "peers"],
  "properties": {
     "topic": {"type": "string"},
     "id": {"type: "string"},
     "subscribed": {"type": "boolean"},
     "peers": {"type": "number", "default": 0}
  }
}
```

I think just the subscribe status and our ID would be good for now. Maybe also the number of peers we have for the topic? If that's annoying to do feel free to exclude it.

> Also I'm thinking Accept: text/event-stream will stream data, and any other header will return JSON.

Works for me!

Reply:
We can do number of peers "we are currently pubsubbing with", but not divided by topic. Let me know if you still want `peers`, and I'll implement the other fields for now.

Reply:
For example it could look like:

`GET /pubsub/example/`

```
{
  "id": "ExampleIDHere",
  "topic": "example",
  "subscribed": false,
  "peers": 0
}
```

Reply:
Feel free to skip peers then. 🤷 It's more of a "nice to have".

Reply:
Ok added code in f313a3f and to spec in d89e712.